### PR TITLE
Migrate ServerError to error_set

### PR DESCRIPTION
**Pull request by haze518**
Replaced thiserror with error_set for ServerError to improve usability and error handling consistency. The transition to error_set not only simplifies the creation of new errors and allows combining them under a common error type, but also provides a convenient mechanism for tracing and logging

### CODE REVIEW

File: server/src/quic/listener.rs [hunk 1]
```diff
@@ -48,7 +48,7 @@ pub fn start(endpoint: Endpoint, system: SharedSystem) {
 async fn handle_connection(
     incoming_connection: quinn::Connecting,
     system: SharedSystem,
-) -> Result<(), ServerError> {
+) -> Result<(), ServerConnectionError> {
```
Comment (numinnex) [inline server/src/quic/listener.rs]: I think we can drop the `Server` prefix from the error name `ConnectionError` should be sufficient and we will use the top level crate name to distinct between server and sdk errors (crates - sdk / iggy )

**Review by numinnex — CHANGES_REQUESTED** [review]

File: server/src/compat/message_conversion/snapshots/retained_batch_snapshot.rs [hunk 2]
```diff
@@ -77,56 +77,56 @@ impl Extendable for RetainedMessageBatchSnapshot {
...
         let base_offset = u64::from_le_bytes(
             value
                 .get(0..8)
-                .ok_or_else(|| {
-                    ServerError::CannotReadMessageBatchFormatConversion(
-                        "Failed to read batch base offset".to_owned(),
-                    )
-                })?
+                .ok_or_else(
+                    || ServerCompatError::CannotReadMessageBatchFormatConversion {
```
Comment (numinnex) [inline server/src/compat/message_conversion/snapshots/retained_batch_snapshot.rs]: I think we can also dip those `reason` fields, those were used as the `context` for our previous error type, now when we will add traces, we will put those in there instead.

File: server/src/quic/listener.rs (hunk 1 above)
Comment (haze518) [reply server/src/quic/listener.rs]: done

File: server/src/compat/message_conversion/snapshots/retained_batch_snapshot.rs (hunk 2 above)
Comment (haze518) [reply server/src/compat/message_conversion/snapshots/retained_batch_snapshot.rs]: yep, I thought about that, but initially decided to leave it as is. However, I like the idea of removing the reason field and adding the trace instead

Comment (haze518) [reply server/src/compat/message_conversion/snapshots/retained_batch_snapshot.rs]: done

**Review by hubcio — APPROVED** [review]

**Pull request merged**